File: test/org/cerion/symcalc/expression/function/SubtractTest.java
package org.cerion.symcalc.expression.function;

import static org.junit.Assert.*;

import org.cerion.symcalc.expression.function.Subtract;
import org.cerion.symcalc.expression.number.IntegerNum;
import org.junit.Test;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.ListExpr;
import org.cerion.symcalc.expression.MathExpr;
import org.cerion.symcalc.expression.VarExpr;

public class SubtractTest {

	@Test
	public void parser() {
		//Verify a-b = Subtract(a,b)
		MathExpr expr = new MathExpr("5-1");
		IntegerNum e = (IntegerNum)expr.eval();
		assertEquals(4, e.toInteger());
	}
	
	@Test
	public void invalidParameters() {
		Expr e = get();
		assertTrue(e.isError());
		
		e = get(new IntegerNum(5));
		assertTrue(e.isError());
		
		e = get(new IntegerNum(1), new IntegerNum(2), new IntegerNum(3));
		assertTrue(e.isError());
		
		//List sizes
		ListExpr a = new IntegerNum(0).toList(2);
		ListExpr b = new IntegerNum(0).toList(3);
		
		e = get(a,b);
		assertTrue(e.isError());
	}
	
	@Test
	public void basicNumbers() {
		
		IntegerNum t = getI(new IntegerNum(5), new IntegerNum(3));
		assertEquals(2, t.toInteger());
		
		t = getI(new IntegerNum(3), new IntegerNum(5));
		assertEquals(-2, t.toInteger());
	}

	@Test
	public void nested() {
		Expr e = new Subtract(new IntegerNum(1), new IntegerNum(2));
		e = new Subtract(new IntegerNum(5), e);

		assertEquals(new IntegerNum(6), e.eval());
	}
	
	@Test
	public void identity() {
		Expr e = get(new VarExpr("a"), new IntegerNum(0));
		
		assertTrue(new VarExpr("a").equals(e));
	}
	
	@Test
	public void lists() {
		
		ListExpr a = new IntegerNum(10).toList(5);
		IntegerNum i = new IntegerNum(3);
		
		//List - Single
		Expr e = get(a,i);
		assertTrue(e.isList());
		
		ListExpr l = (ListExpr)e;
		assertEquals(5, l.size());
		
		for(int n = 0; n < 5; n++)
			assertEquals(7, ((IntegerNum)l.get(n)).toInteger());
		
		//Single - List
		e = get(i,a);
		
		l = (ListExpr)e;
		assertEquals(5, l.size());
		
		for(int n = 0; n < 5; n++)
			assertEquals(-7, ((IntegerNum)l.get(n)).toInteger());
	}

	
	private Expr get(Expr ...e)
	{
		return new Subtract(e).eval();
	}
	
	private IntegerNum getI(Expr ...e)
	{
		Expr t = new Subtract(e).eval();
		
		return (IntegerNum)t;
	}
}

File: symcalc/src/test/java/function/arithmetic/SubtractTest.kt
package org.cerion.symcalc.function.arithmetic

import org.cerion.symcalc.`==`
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.Expr.ExprType
import org.cerion.symcalc.expression.ListExpr
import org.cerion.symcalc.expression.VarExpr
import org.cerion.symcalc.number.Integer
import kotlin.test.Test
import kotlin.test.assertEquals

class SubtractTest {

    @Test
    fun parser() {
        val expr = Expr.parse("5-1")
        assertEquals(Integer(4), expr.eval())
    }

    @Test
    fun invalidParameters() {
        assertEquals(ExprType.ERROR, Subtract().eval().type)
        assertEquals(ExprType.ERROR, Subtract(Integer(5)).eval().type)
        assertEquals(ExprType.ERROR, Subtract(Integer(1), Integer(2), Integer(3)).eval().type)
    }

    @Test
    fun basicNumbers() {
        assertEquals(Integer(2), Subtract(Integer(5), Integer(3)).eval())
        assertEquals(Integer(-2), Subtract(Integer(3), Integer(5)).eval())
    }

    @Test
    fun nested() {
        var e: Expr = Subtract(Integer(1), Integer(2))
        e = Subtract(Integer(5), e)

        assertEquals(Integer(6), e.eval())
    }

    @Test
    fun identity() {
        Subtract("a", 0) `==` VarExpr("a")
        Subtract(0, "a") `==` Times(-1, VarExpr("a"))
    }

    @Test
    fun lists() {
        Subtract(ListExpr(1,5,8), ListExpr(7,2,5)) `==` ListExpr(-6,3,3)
        Subtract(ListExpr(3,4,5), 2) `==` ListExpr(1,2,3)
        Subtract(10, ListExpr(3,4,5)) `==` ListExpr(7,6,5)
    }
}
